simpleflow/execute.py

```diff
@@ -0,0 +1,193 @@
+    # func() or func(**kwargs) or func(a=1, b=2)
+    if not argspec.varargs and not argspec.args and args:
+        raise TypeError('command does not take varargs')
+
+    # Calling func(a, b) with func(1, 2, 3)
+    if (not argspec.varargs and argspec.args and
+            len(args) != len(argspec.args)):
+        raise TypeError('command takes {} arguments: {} passed'.format(
+            len(argspec.args),
+            len(args)))
+
+
+def check_keyword_arguments(argspec, kwargs):
+    # func() or func(*args) or func(a, b)
+    if not argspec.keywords and not argspec.defaults and kwargs:
+        raise TypeError('command does not take keyword arguments')
+
+    arguments_defaults = zip_arguments_defaults(argspec)
+    not_found = (set(name for name, value in arguments_defaults if
+                     value is RequiredArgument) -
+                 set(kwargs))
+    # Calling func(a=1, b) with func(2) instead of func(a=0, 2)
+    if not_found:
+        raise TypeError('argument{} "{}" not found'.format(
+            's' if len(not_found) > 1 else '',
+            ', '.join(not_found)))
+
+
+def format_arguments_json(*args, **kwargs):
+    return json.dumps({
+        'args': args,
+        'kwargs': kwargs,
+    })
+
+
+def get_name(func):
+    import types
+
+    prefix = func.__module__
+
+    if not callable(func):
+        raise ValueError('{} is not callable'.format(
+            func))
+
+    if hasattr(func, 'name'):
+        name = func.name
+    elif isinstance(func, types.FunctionType):
+        name = func.func_name
+    else:
+        name = func.__class__.__name__
+
+    return '.'.join([prefix, name])
+
+
+def python(interpreter='python'):
+    """
+    Execute a callable as an external Python program.
+
+    One of the use cases is to use a different interpreter than the current one
+    such as pypy.
+
+    Arguments of the decorated callable must be serializable in JSON.
+
+    """
+    def wrap_callable(func):
+        @functools.wraps(func)
+        def execute(*args, **kwargs):
+            command = 'simpleflow.execute'  # name of a module.
+            try:
+                output = subprocess.check_output([
+                    interpreter, '-m', command,  # execute module a script.
+                    get_name(func), format_arguments_json(*args, **kwargs),
+                ],
+                    # Redirect stderr to stdout to get traceback on error.
+                    stderr=subprocess.STDOUT,
+                )
+            except subprocess.CalledProcessError as err:
+                print err.output
+                excline = err.output.rsplit('\n', 2)[1]
+                try:
+                    excname, excmsg = excline.split(':', 1)
+                except ValueError:
+                    excname = excline
+                    excmsg = ''
+
+                exception = make_callable(excname.strip())
+                raise exception(excmsg.strip())
+            return json.loads(output)
```

Comment:
I don't understand how this part work, `output` is a byte string, so we assume that the command is returning serialized JSON ?